pkg/runtime/scheme.go

```diff
@@ -75,111 +130,218 @@ func NewScheme() *Scheme {
 func (s *Scheme) KnownTypes(gv unversioned.GroupVersion) map[string]reflect.Type {
-	return s.raw.KnownTypes(gv)
+	types := make(map[string]reflect.Type)
+	for gvk, t := range s.gvkToType {
+		if gv != gvk.GroupVersion() {
+			continue
+		}
+
+		types[gvk.Kind] = t
+	}
+	return types
 }
 
-// ObjectKind returns the default group,version,kind of the given Object.
+// ObjectKind returns the group,version,kind of the go object,
+// or an error if it's not a pointer or is unregistered.
 func (s *Scheme) ObjectKind(obj Object) (unversioned.GroupVersionKind, error) {
-	return s.raw.ObjectKind(obj)
+	gvks, err := s.ObjectKinds(obj)
+	if err != nil {
+		return unversioned.GroupVersionKind{}, err
+	}
+	return gvks[0], nil
 }
 
-// ObjectKinds returns the all possible group,version,kind of the given Object.
+// ObjectKinds returns all possible group,version,kind of the go object,
+// or an error if it's not a pointer or is unregistered.
 func (s *Scheme) ObjectKinds(obj Object) ([]unversioned.GroupVersionKind, error) {
-	return s.raw.ObjectKinds(obj)
+	v, err := conversion.EnforcePtr(obj)
+	if err != nil {
+		return nil, err
+	}
+	t := v.Type()
+
+	gvks, ok := s.typeToGVK[t]
+	if !ok {
+		return nil, &notRegisteredErr{t: t}
+	}
+
+	return gvks, nil
 }
 
 // Recognizes returns true if the scheme is able to handle the provided group,version,kind
 // of an object.
 func (s *Scheme) Recognizes(gvk unversioned.GroupVersionKind) bool {
-	return s.raw.Recognizes(gvk)
+	_, exists := s.gvkToType[gvk]
+	return exists
 }
 
 func (s *Scheme) IsUnversioned(obj Object) (bool, bool) {
-	return s.raw.IsUnversioned(obj)
+	v, err := conversion.EnforcePtr(obj)
+	if err != nil {
+		return false, false
+	}
+	t := v.Type()
+
+	if _, ok := s.typeToGVK[t]; !ok {
+		return false, false
+	}
+	_, ok := s.unversionedTypes[t]
+	return ok, true
 }
 
 // New returns a new API object of the given version ("" for internal
 // representation) and name, or an error if it hasn't been registered.
 func (s *Scheme) New(kind unversioned.GroupVersionKind) (Object, error) {
-	obj, err := s.raw.NewObject(kind)
-	if err != nil {
-		return nil, err
+	if t, exists := s.gvkToType[kind]; exists {
+		return reflect.New(t).Interface().(Object), nil
+	}
+
+	if t, exists := s.unversionedKinds[kind.Kind]; exists {
+		return reflect.New(t).Interface().(Object), nil
 	}
-	return obj.(Object), nil
+	return nil, &notRegisteredErr{gvk: kind}
 }
 
 // Log sets a logger on the scheme. For test purposes only
 func (s *Scheme) Log(l conversion.DebugLogger) {
-	s.raw.Log(l)
+	s.converter.Debug = l
 }
 
-// AddConversionFuncs adds a function to the list of conversion functions. The given
-// function should know how to convert between two API objects. We deduce how to call
-// it from the types of its two parameters; see the comment for
-// Converter.RegisterConversionFunction.
+// AddIgnoredConversionType declares a particular conversion that should be ignored - during conversion
+// this method is not invoked.
```

Comment:
Can you clarify this? What does ignored mean? I think the comment from the version of this function that you're deleting is clearer.